dev/Microsoft.Process.Environment/EnvironmentManager.cpp

```diff
@@ -141,12 +136,157 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+        std::wstring existingPath = GetPath();
+
+        // Don't append to the path if the addition already exists.
+        if (existingPath.find(path) != std::wstring::npos)
+        {
+            return;
+        }
+
+        std::wstring newPath = existingPath.append(path);
+
+        if (newPath.back() != L';')
+        {
+            newPath += L';';
+        }
+
+        auto setPath = [&, newPath, this]()
+        {
+            if (m_Scope == Scope::Process)
+            {
+                BOOL result = FALSE;
+                result = ::SetEnvironmentVariable(L"Path", newPath.c_str());
+
+                if (result == 0)
+                {
+                    THROW_HR(HRESULT_FROM_WIN32(GetLastError()));
+                }
+            }
+            else //Scope is either user or machine
+            {
+                wil::unique_hkey environmentVariableKey = GetRegHKeyForEVUserAndMachineScope(true);
+
+                THROW_IF_FAILED(HRESULT_FROM_WIN32(RegSetValueEx(
+                    environmentVariableKey.get()
+                    , L"Path"
+                    , 0
+                    , REG_EXPAND_SZ
+                    , reinterpret_cast<const BYTE*>(newPath.c_str())
+                    , static_cast<DWORD>((newPath.size() + 1) * sizeof(wchar_t)))));
+
+                LRESULT broadcastResult = SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE,
+                    reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
+                    SMTO_NOTIMEOUTIFNOTHUNG | SMTO_BLOCK, 1000, nullptr);
+
+                if (broadcastResult == 0)
+                {
+                    THROW_HR(HRESULT_FROM_WIN32(GetLastError()));
+                }
+            }
+
+            return S_OK;
+        };
+
+        EnvironmentVariableChangeTracker changeTracker(L"Path", std::wstring(newPath), m_Scope);
+
+        THROW_IF_FAILED(changeTracker.TrackChange(setPath));
     }
 
     void EnvironmentManager::RemoveFromPath(hstring const& path)
     {
-        throw hresult_not_implemented();
+        if (path.empty())
+        {
+            THROW_HR(E_INVALIDARG);
+        }
+
+        // A user is only allowed to remove something from the PATH if
+        // 1. path exists in PATH
+        // 2. path matches a path part exactly (ignoring case)
+        std::wstring currentPath = GetPath();
+        std::wstring pathPartToFind(path);
+
+        if (pathPartToFind.back() != L';')
+        {
+            pathPartToFind += L';';
+        }
+
+        int left = static_cast<int>(currentPath.size());
+        int right = static_cast<int>(currentPath.size() - 1);
+        bool foundPathPart = false;
+
+        while (!foundPathPart && left >= 0)
+        {
+            if (left == 0 || currentPath[left - 1] == L';')
+            {
+                std::wstring pathPart(currentPath, left, right - left);
+
+                if (_wcsnicmp(pathPart.c_str(), pathPartToFind.c_str(), pathPart.size() + 1) == 0)
```

Comment:
@DrusTheAxe  I think you're confusing the infamous l* comparison functions with wcscmp here, which performs an ordinal comparison. https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strcmp-wcscmp-mbscmp

Reply:
Oddly enough I see email about a comment by @riverar but don't see the actual comment in the github view

>@DrusTheAxe I think you're confusing the infamous l* comparison functions with wcscmp here, which performs an ordinal comparison. https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strcmp-wcscmp-mbscmp

Yes the lstrcmp/etc functions are wrappers over CompareString etc lexical compares (https://docs.microsoft.com/windows/win32/intl/handling-sorting-in-your-applications) but I put those out of mind years ago :P

I was thinking of good old CRT wcs*(). Maybe I'm thinking of wcs* behavior varying dependent on locale? [][Sort Strings Ordinally](https://docs.microsoft.com/windows/win32/intl/handling-sorting-in-your-applications#sort-strings-ordinally) is pretty blunt about 'use CompareStringOrdinal' but vague on if wcs* are equivalent (and perhaps maybe, having changed over time?). Internally we've long had strong guidance "only use CompareStringOrdinal. Don't use wcs*" when handling package identity, application identity, filesystem, registry keys, and other ordinal-y data but perhaps things have changed?

@ShawnSteele are wcs* compare functions guaranteed to perform ordinal comparisons equivalent to `CompareStringOrdinal`?

Reply:
I quickly retracted it, because the behavior for the wide functions, when dealing with locale, gets really muddy and I think your original prescriptive guidance--replace the sucker!--is best. 😀

Reply:
This isn't the first time I've ran into this issue.  However, the epiphany just came to me.

Ordinal is binary.  Compares the bits.  This is when we want that sweet, sweet precisision.

Lexical, is based on the letters and numbers.  The out put, <0, 0, >0 is still the same, but what is compared is slightly different.